package/MDAnalysis/core/topologyattrs.py

```diff
@@ -801,9 +824,50 @@ def center_of_mass(group, pbc=None):
 
 
         .. versionchanged:: 0.8 Added `pbc` parameter
+        .. versionchanged:: 0.18.0 Added `compound` parameter
         """
-        return group.atoms.center(weights=group.atoms.masses,
-                                  pbc=pbc)
+        atoms = group.atoms.unique
+        # COM of the whole group:
+        if compound == 'group':
+            return atoms.center(weights=atoms.masses, pbc=pbc)
+        # If COMs per compound are requested, get their number and indices:
+        elif compound.lower() == 'residues':
+            compound_indices = atoms.resindices
+            n_compounds = atoms.n_residues
+        elif compound.lower() == 'segments':
+            compound_indices = atoms.segindices
+            n_compounds = atoms.n_segments
+        else:
+            raise ValueError("Unrecognized compound definition: {}\nPlease use"
+                             " one of 'group', 'residues', or 'segments'."
+                             "".format(compound))
+        # Sort positions and masses by compound index:
+        sort_indices = np.argsort(compound_indices)
+        compound_indices = compound_indices[sort_indices]
+        positions = atoms.positions[sort_indices]
+        masses = atoms.masses[sort_indices]
+        # Allocate output array:
+        coms = np.zeros((n_compounds, 3), dtype=np.float32)
+        # Get sizes of compounds:
+        unique_compound_indices, compound_sizes = np.unique(compound_indices,
+                                                            return_counts=True)
+        unique_compound_sizes = np.unique(compound_sizes)
+        # Compute COMs per compound for each compound size:
```

Comment:
So if I'm reading this right, to reduce the number of numpy calls, you group together similar sized residues and do all 6 atom residue COM calculations together?

Reply:
Exactly!